signature/sign_test.go

```diff
@@ -134,6 +139,126 @@ func TestSignVerify(t *testing.T) {
 	}
 }
 
+var _ crypto.Signer = MockCryptoSigner{}
+
+type MockCryptoSigner struct {
+	privateKey crypto.PrivateKey
+	publickKey crypto.PublicKey
+}
+
+func (m MockCryptoSigner) Public() crypto.PublicKey { return m.publickKey }
+
+func (m MockCryptoSigner) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+	return ecdsa.SignASN1(rand, m.privateKey.(*ecdsa.PrivateKey), digest)
+}
+
+func (m MockCryptoSigner) Algorithm() jwa.KeyAlgorithm {
+	return jwa.ES256
+}
+
+func TestSignVerifyCryptoSigner(t *testing.T) {
+
+	t.Parallel()
+	ctx := context.Background()
+
+	step := &pipeline.CommandStep{
+		Command: "llamas",
+		Plugins: pipeline.Plugins{
+			{
+				Source: "some-plugin#v1.0.0",
+				Config: nil,
+			},
+			{
+				Source: "another-plugin#v3.4.5",
+				Config: map[string]any{"llama": "Kuzco"},
+			},
+		},
+		Env: map[string]string{
+			"CONTEXT": "cats",
+			"DEPLOY":  "0",
+		},
+	}
+	// The pipeline-level env that the agent uploads:
+	signEnv := map[string]string{
+		"DEPLOY": "1",
+	}
+
+	// The backend combines the pipeline and step envs, providing a new env:
+	verifyEnv := map[string]string{
+		"CONTEXT": "cats",
+		"DEPLOY":  "0",
+		"MISC":    "llama drama",
+	}
+
+	stepWithInvariants := &CommandStepWithInvariants{
+		CommandStep:   *step,
+		RepositoryURL: "fake-repo",
+	}
+
+	cases := []struct {
+		name              string
+		alg               jwa.SignatureAlgorithm
+		expectedSignature string
+	}{
+		{
+			name:              "should sign using crypto.Signer",
+			alg:               jwa.ES256,
+			expectedSignature: "eyJhbGciOiJFUzI1NiJ9..Op5KSww95n5s1b9jz0Me5UGqUQPcHzEIFvkWTB_yEv6qEDnnFUO1XsC5592fQoAcB0VnPnHaK31iSiCypREIdA",
+		},
+	}
+
+	wd, err := os.Getwd()
+	if err != nil {
+		t.Fatalf("os.Getwd() error = %v", err)
+	}
+
+	privateKeyPath := path.Join(wd, "fixtures", "crypto_signer", "P256", "private.pem")
+	pemPrivateKey, err := os.ReadFile(privateKeyPath)
+	if err != nil {
+		t.Fatalf("os.ReadFile(%q) error = %v", privateKeyPath, err)
+	}
+
+	block, _ := pem.Decode([]byte(pemPrivateKey))
```

Comment:
Funny how this _looks_ like an ignored error, but isn't. According to the docs, `block` would be nil if it failed to parse. Arguably we could check that and `t.Fatalf`, but since the following line would panic, it's not a big deal.